Add support for Python 3.8 / C-based pickle

mwtian opened the pull request:
In Python 3.8, C-based `pickle` supporting protocol 5 becomes the default. `pickle`'s APIs to write custom serializers also changed. This PR adds support for the new `pickle` API. Tests pass under Python 3.8, except `test_cloudpickle_generators.py::test_self_in_closure`, which I'm not totally sure how it works.

This also works with Python 3.9. In Python 3.10, there is a C compile error which needs to be investigated separately.

Commit:
support c-based pickle

llllllllll commented on cloudpickle_generators/__init__.py:
```diff
@@ -221,29 +221,118 @@ def _save_generator_impl(self, frame, gen, filler):
+        frame.f_globals,
+        gen.__name__,
+        (),
+        (_empty_cell(), ) * len(f_code.co_freevars),
```
this needs to to include the cell vars also.

EDIT: actually, I think I am just misremembering how the function constructor works, the cell var cells cells live in the locals of the frame, my bad.

llllllllll commented on cloudpickle_generators/__init__.py:
```diff
@@ -221,29 +221,118 @@ def _save_generator_impl(self, frame, gen, filler):
-        CloudPickler.dispatch[CoroutineType] = _save_coroutine
-    if sys.version_info >= (3, 6, 0):
-        CloudPickler.dispatch[AsyncGeneratorType] = _save_async_generator
+    if pickle.HIGHEST_PROTOCOL >= 5:
```
should this be branching based on the highest protocol or the actual protocol being requested at the time of serialization?

llllllllll commented on cloudpickle_generators/tests/test_cloudpickle_generators.py:
```diff
@@ -124,6 +125,7 @@ def g():
     g(0)
 
 
+@pytest.mark.xfail(sys.version_info >= (3, 8, 0), reason="needs to be investigated")
```
Whats the error here? This is an important test for the skeleton + fill pattern.

The point of this test is to make a recursive structure where one of the free variables of a generator is itself. The point of the skeleton is that it is flat, so that when we can serialize it and then fill it back up once we are done deserializing everything.

llllllllll commented on cloudpickle_generators/__init__.py:
```diff
@@ -221,29 +221,118 @@ def _save_generator_impl(self, frame, gen, filler):
+
+
+def _reduce_generator(gen):
+    if gen.gi_running:
+        raise ValueError('cannot save running generator')
```
we need to have the equivalent check for the other reduce functions

llllllllll reviewed (commented):
Thanks for adding support for more modern Python and pickle!

One style comment on the `_reduce` and `_rehydrate` pairs. I spent a minute trying to figure out what, if anything, was different between them. It might be helpful to create a builder function which takes the name of the attribute frame attribute and returns the pair of functions.

mwtian commented:
I'm still working on addressing the comments, especially fixing `test_self_in_closure()`. The error is `NameError: free variable 'nop' referenced before assignment in enclosing scope`. It looks like captured function is not handled correctly.
- For nested function like the test, `nop` appears in `f_locals` of the frame before serialization / deserialization roundtrip, but disappears after the roundtrip.
- When the functions are in global scope, serialization hits max recursion depth error.

And thanks for the useful and magical library! I'm experimenting with integrating this library into [Ray](https://github.com/ray-project/ray), to allow describing persistent workflow or RPC flow via Python coroutines: https://github.com/ray-project/ray/issues/21793. Welcome your suggestions.
